core/trino-main/src/main/java/io/trino/execution/scheduler/HashDistributionSplitAssigner.java

```diff
@@ -93,33 +120,43 @@ public AssignmentResult assign(PlanNodeId planNodeId, ListMultimap<Integer, Spli
             }
         }
         else {
-            for (Integer outputPartitionId : splits.keySet()) {
+            splits.forEach((outputPartitionId, split) -> {
                 TaskPartition taskPartition = outputPartitionToTaskPartition.get(outputPartitionId);
                 verify(taskPartition != null, "taskPartition not found for outputPartitionId: %s", outputPartitionId);
-                if (!taskPartition.isIdAssigned()) {
-                    // Assigns lazily to ensure task ids are incremental and with no gaps.
-                    // Gaps can occur when scanning over a bucketed table as some buckets may contain no data.
-                    taskPartition.assignId(nextTaskPartitionId++);
+
+                List<SubPartition> subPartitions;
+                if (taskPartition.getSplitBy().isPresent() && taskPartition.getSplitBy().get().equals(planNodeId)) {
+                    subPartitions = ImmutableList.of(taskPartition.getNextSubPartition());
                 }
-                int taskPartitionId = taskPartition.getId();
-                if (!createdTaskPartitions.contains(taskPartitionId)) {
-                    Set<HostAddress> hostRequirement = sourcePartitioningScheme.getNodeRequirement(outputPartitionId)
-                            .map(InternalNode::getHostAndPort)
-                            .map(ImmutableSet::of)
-                            .orElse(ImmutableSet.of());
-                    assignment.addPartition(new Partition(
-                            taskPartitionId,
-                            new NodeRequirements(catalogRequirement, hostRequirement)));
-                    for (PlanNodeId replicatedSource : replicatedSplits.keySet()) {
-                        assignment.updatePartition(new PartitionUpdate(taskPartitionId, replicatedSource, replicatedSplits.get(replicatedSource), completedSources.contains(replicatedSource)));
-                    }
-                    for (PlanNodeId completedSource : completedSources) {
-                        assignment.updatePartition(new PartitionUpdate(taskPartitionId, completedSource, ImmutableList.of(), true));
+                else {
+                    subPartitions = taskPartition.getSubPartitions();
```

Comment:
is `subPartitions` always a singleton here?

Reply:
Yeah, it is enforced in the `TaskPartition` constructor